### ISSUE TITLE
Image rotation

### ISSUE THREAD

**Issue by lstrzebinczyk** [ruby2d/ruby2d#30]
Hi again!

Would it be possible to change rotation of images? In my example I would like to rotate image of a character 90 degrees right to indicate that the character is sleeping.

I don't see any code I would be able to use in current code, maybe this would be easy to add with some additions to the c layer?

**Comment by blacktm** [comment]
Anything is possible! 🔮
But sadly, this hasn't been implemented yet 😢

I see this as being pretty essential — actually, rotation of all drawable objects that is, but textures are a good place to start. I think it's best to push the actual rendering implementation to a dependency, the [Simple 2D project](https://github.com/simple2d), for both sharing and efficiency. But that still leaves some decisions on what the Ruby 2D API should look like. A simple `rotate` method that takes an angle in degrees? Does it rotate around the center point, or another anchor? Does this then change the x, y coordinates? Open to ideas here!

**Comment by KillaPL** [comment]
For my uses rotation around center is good, but I can easily imagine needing to rotate around any point.

I think something like this would be adequate:

```
image = Image.new(...)
image.rotate(around: orientation_center, radians: radians_value, degrees: degrees_value) 
image.reset_rotation
```

with the following rules:

* `orientation_center` would be one of: `:center` or a hash with :x and :y.
* `radians_value` must be a number
* `degrees_value` must be a number
* only `radians_value` or `degrees_value` can be provided at the same time

And of course image is rotated according to those parameters, geometrically this is pretty simple.

`Image#reset_orientation`, unsurprisingly, cancell orientation data. I'm not sure if this is really needed.

What do you think?

**Comment by KillaPL** [comment]
Or maybe simply adding `rotation` as additional parameter, one that would define rotation between images top from x axis?

**Comment by blacktm** [comment]
All good ideas here. Maybe we could start with a simple `rotate` param as you mentioned for a starting point. I could see us implementing this is the Simple 2D project also, so others can use it at a lower level. Any ideas about the math here? Could do a [`glRotate`](https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glRotate.xml) here, but that's only in OpenGL 2 — would have to do something else for GL3 and GLES.

**Comment by KillaPL** [comment]
![rotate](https://cloud.githubusercontent.com/assets/1272828/25799862/14f1e84e-33e7-11e7-8ec4-6df21b6ad3a5.png)

This is how I see it, in a minimal way. Rotate should be a float value to represent rotation of given point around x axis. I'm not sure if this should be around that x0, y0 point or the middle. Maybe the middle will be easier to handle in the end...

**Comment by blacktm** [comment]
Ha, what are you talking about — that looks great!

Ok, been thinking about all this. We can design the Ruby interface to be whatever we want, no problem there. The more challenging piece is how we actually implement rotation of textures. I've looked around for the best place to do this, and [I think it's here](https://github.com/simple2d/simple2d/blob/master/src/gl3.c#L201), where we would simply manipulate its vertices right before it's drawn (this is in GL3, but it's nearly identical to GL2, GLES, WebGL). I don't have the math worked out yet, but I played around with these vertices just to verify the approach, and it seems to have the desired result:

![screen shot 2017-05-19 at 10 39 58 pm](https://cloud.githubusercontent.com/assets/887429/26272402/2996961c-3ce5-11e7-81b2-a4db5825a8f3.png)

What do you think, should we go for it?

**Comment by KillaPL** [comment]
The edges look very jaggedy, is this by design? Can we have antyaliasing in easily achievable manner?

**Comment by blacktm** [comment]
Boom! 💥

![screen shot 2017-05-20 at 11 07 22 pm](https://cloud.githubusercontent.com/assets/887429/26280843/30c34c88-3db1-11e7-85a6-3f343942bf2c.png)

Actually isn't that hard to add antialiasing to everything rendered (shapes too), just have to add...
```
SDL_GL_SetAttribute(SDL_GL_MULTISAMPLEBUFFERS,  1);
SDL_GL_SetAttribute(SDL_GL_MULTISAMPLESAMPLES, 16);
```
...somewhere in `S2D_GL_Init`. And ignore the PNG — the pixels inside textures aren't touched (it's part of the image data, not an edge), although we do have the ability to blend those separately. It actually looks more pronounced here as well because I'm on a Retina MacBook and we don't currently have HiDPI enabled for OpenGL (only WebGL right now because browsers are awesome and make it effortless), hence why the title bar text looks sharp and images were jagged in the original screen shot. Maybe we can add a feature separately to enable antialiasing and set the number of samples.

Getting back to rotating image vertices, since you're a math wiz 😉, any ideas of where to start? Is there a specific mathematical approach you'd suggest to take, or begin researching? Ideally, we would be able to rotate the four `x, y` vertices around any arbitrary point.

**Comment by blacktm** [comment]
Implemented in 6acf6c8bd0cd8051bdb6cb6ec75fb7d67959e47b 🌀(and in https://github.com/simple2d/simple2d/commit/ea8a0a031144cddf385237bebe920079a8ac6c62).
